README
# de.tuxteam.cms.notify

Schickt bei Änderungen an der DAV-Datenbank eine Nachricht in einen Jabber Multi-User-Chat.


## Usage

`java -jar cms-notify.jar config.ini`

Beispielkonfiguration:

```
    [jabber]
    host = cms-backend.staging.zeit.de
    user = cms-backend
    password = secret
    muc = notifications@conference.cms-backend.staging.zeit.de
    muc-password =

    [postgres]
    url = //localhost:5432/cms
    user = cms-reader
    password =
```

Nach Installation als deb gibt es ein `/etc/init.d/cms-notify`, das die
Konfiguration in `/etc/cms-notify.conf` erwartet.

XXX Loggen tut es leider bisher nichts, bei Problemen also am besten mal im
Vordergrund starten.


## Building

deb-Paket erstellen in Vagrant-VM mit `./build-scripts/make-deb.sh`

Lokal bauen mit `lein uberjar` (dazu http://leiningen.org/ installieren)


## Funktionsweise

Das Apache-Modul mod_dav_cms schreibt bei jeder Änderung, die es durchführt,
zusätzlich eine Zeile in die Tabelle `triggers`. Die lesen wir hier einfach
periodisch aus und übersetzen ihre Zeilen in Jabber-Nachrichten.


debian/changelog
cms-notify (1.1.0-zeit1) unstable; urgency=medium

  * Transmit timestamp so it works with timezone information

 -- Wolfgang Schnerring <wolfgang.schnerring@zeit.de>  Thu, 27 Apr 2017 13:07:53 +0200

cms-notify (1.0.0-zeit2) unstable; urgency=medium

  * Compress deb with gzip so it works with Ubuntu-10

 -- Wolfgang Schnerring <wolfgang.schnerring@zeit.de>  Mon, 05 Sep 2016 12:07:53 +0200

cms-notify (1.0.0-zeit1) unstable; urgency=medium

  * Initial packaging

 -- Wolfgang Schnerring <zon-backend@zeit.de>  Thu, 01 Sep 2016 15:29:09 +0200


test/de/tuxteam/cms/notify/test/core.clj
(ns de.tuxteam.cms.notify.test.core
  (:use [de.tuxteam.cms.notify.core])
  (:use [clojure.test]))

(deftest replace-me ;; FIXME: write
  (is false "No tests have been written."))


src/de/tuxteam/cms/notify/core.clj
;; Notify

(ns de.tuxteam.cms.notify.core
	(:gen-class)
        (:use clojure.contrib.sql)
        (:use clojure-ini.core)
        (import (java.util Date))
        (import (java.sql Timestamp))
        (import (java.text DateFormat))
        (import (org.jivesoftware.smack ConnectionConfiguration
                                        SASLAuthentication
                                        XMPPConnection
                                        XMPPException))
        (import (org.jivesoftware.smack.packet Message Message$Type))
        (import (org.jivesoftware.smackx.muc MultiUserChat))
        (import (org.jivesoftware.smackx XHTMLManager XHTMLText)))

;;;;############################################################################
;;;;                                                           Globals
;;;;############################################################################

;;; A list of resource events that we want to report
(def interesting-events (hash-set "PROPPATCH" "MKCOL" "PUT" "DELETE" "MOVE"))

;;; Our connection to the jabber server
(def *jabber-connection* nil)

;;; The conference room
(def *conference* nil)

;;; How long should we sleep inbetween polls? (in ms.)
(def *sleep-time* 20000)

;;;;############################################################################
;;;;                                                      Mutable Data
;;;;############################################################################

;;; `last-event' holds the timestamp of the last time we looked at the
;;; event table. We initialize it to the current time at
;;; startup. During a poll to the database we mutate it to the
;;; transaction time of the query.  We might be able to get away
;;; without this by passing the last seen time to the next poll
;;; call. Maybe next time ;-/

(def last-event (ref (Timestamp. (.getTime (Date.)))))

;;; Should we run/sleep/stop?
(def server-state (ref :running))


;;;;############################################################################
;;;;                                                      Code Section
;;;;############################################################################

(defn patch-jabber-library
  "Dissable over-correct SAL methods for now"
  []
  (doseq [method ["EXTERNAL" "GSSAPI" "DIGEST-MD5" "CRAM-MD5"]]
    (SASLAuthentication/unregisterSASLMechanism method))
  (SASLAuthentication/supportSASLMechanism "PLAIN" 0))

(defn init-jabber-connection [config]
  "Initialize the connection to the jabber server"
  (let [conn (new XMPPConnection (:host config))]

    (doto conn
      (.connect)
      (.login (:user config) (:password config)))
    (def *jabber-connection* conn)))

(defn join-conference [config]
  (let [muc (MultiUserChat. *jabber-connection* (:muc config))]
    (.join muc (:user config) (:muc-password config))
    (def *conference* muc)))

(defn send-notification [text]
  (let [message (.createMessage *conference*)]
    ;; FIXME: we probably need to create more structured content here ...
    (doto message
      (.setBody (str "Resource changed: " text)))
    (.sendMessage *conference* message)))

(defn send-hello []
  (let [message (.createMessage *conference*)]
    ;; FIXME: we probably need to create more structured content here ...
    (doto message
      (.setBody (str "Hello from CMS-Backend")))
      (.sendMessage *conference* message)))

(defn process-interesting-events
  [event]
  (when (contains?  interesting-events (:method event))
    (send-notification (:source event))))

(defn check-resources
  ""
  []
  (println "Checking for changed resource since" @last-event)
  (transaction
   (with-query-results rs
     ["SELECT distinct(source), method, now() AS now FROM triggers WHERE logdate >= ?"  @last-event]
     (when (first rs)
       (dosync (ref-set last-event (:now (first rs)))))
     (if (first rs)
       (println "Have results") (println "No results"))
     (dorun (map #(process-interesting-events %) (into (hash-set)  rs))))))



(defn run-event-loop
  ""
  [config]
  (def database {:classname   "org.postgresql.Driver" ; must be in classpath
                 :subprotocol "postgresql"
                 :subname     (:url config)
                 :user        (:user config)
                 :password    (:password config)})
  (with-connection database
    (println "Connected to database")
    (while (= @server-state :running)
      (check-resources)
      (Thread/sleep *sleep-time*))))	


;;;;############################################################[ Main ]########

(defn -main [configfile]
  (def config (read-ini configfile :keywordize? true :comment-char \#))
  (patch-jabber-library)
  (init-jabber-connection (:jabber config))
  (join-conference (:jabber config))
  (run-event-loop (:postgres config)))


src/de/tuxteam/cms/notify/old_core.clj
;; Notify

(ns de.tuxteam.cms.Notify
	(:gen-class)
	(use clojure.contrib.repl-ln)
  (:use clojure.contrib.sql)
  (import (org.jivesoftware.smack ConnectionConfiguration 
                                  SASLAuthentication
                                  XMPPConnection 
                                  XMPPException))
  (import (java.util Date))
  (import (java.sql Timestamp))
  (import (java.text DateFormat))
  (import (org.jivesoftware.smack.packet Message Message$Type))
  (import (org.jivesoftware.smackx.muc MultiUserChat))
  (import (org.jivesoftware.smackx XHTMLManager XHTMLText)))

;;;;############################################################################################
;;;;                                                                     Customization
;;;;############################################################################################

;;;;###################################################################[ Jabber Server ]########

(def jabber-host     "zip6.zeit.de")
(def jabber-user     "cms-backend")
(def jabber-password "--arag0rn+")
(def conference-room "notifications@conference.zip6.zeit.de")
(def muc-password    "")
(def cms-room        "notifications")

;;;;###############################################################[ Database Settings ]########

(def database {:classname   "org.postgresql.Driver" ; must be in classpath
               :subprotocol "postgresql"
               :subname      (str "//" "localhost" ":" 5432 "/" "cms") ;"//localhost:5432/cms"
               :user        "cms-reader"
               :password    ""})


;;;;############################################################################################
;;;;                                                                           Globals
;;;;############################################################################################

;;; A list of resource events that we want to report
(def interesting-events (hash-set "PROPPATCH" "MKCOL" "PUT" "DELETE" "MOVE"))

;;; Our connection to the jabber server
(def *jabber-connection* nil)

;;; The conference room
(def *conference* nil)

;;; How long should we sleep inbetween polls? (in ms.)
(def *sleep-time* 20000)

;;;;############################################################################################
;;;;                                                                      Mutable Data
;;;;############################################################################################

;;; `last-event' holds the timestamp of the last time we looked at the
;;; event table. We initialize it to the current time at
;;; startup. During a poll to the database we mutate it to the
;;; transaction time of the query.  We might be able to get away
;;; without this by passing the last seen time to the next poll
;;; call. Maybe next time ;-/

(def last-event (ref (str (Timestamp. (.getTime (Date.))))))

;;; Should we run/sleep/stop?
(def server-state (ref :running))


;;;;############################################################################################
;;;;                                                                      Code Section
;;;;############################################################################################

(defn patch-jabber-library 
  "Dissable over-correct SAL methods for now"
  []
  (doseq [method ["EXTERNAL" "GSSAPI" "DIGEST-MD5" "CRAM-MD5"]]
    (SASLAuthentication/unregisterSASLMechanism method))
  (SASLAuthentication/supportSASLMechanism "PLAIN" 0))

(defn init-jabber-connection []
  "Initialize the connection to the jabber server"
  (let [conn (new XMPPConnection jabber-host)]
    
    (doto conn  
      (.connect)
      (.login jabber-user jabber-password))
    (def *jabber-connection* conn)))

(defn join-conference []
  (let [muc (MultiUserChat. *jabber-connection* conference-room)]
    (.join muc jabber-user muc-password)
    (def *conference* muc)))

(defn send-notification [text]
  (let [message (.createMessage *conference*)]
    ;; FIXME: we probably need to create more structured content here ...
    (doto message
      (.setBody (str "Resource changed: " text)))
    (.sendMessage *conference* message)))

(defn send-hello []
  (let [message (.createMessage *conference*)]
    ;; FIXME: we probably need to create more structured content here ...
    (doto message
      (.setBody (str "Hello from CMS-Backend")))
      (.sendMessage *conference* message)))

(defn process-interesting-events 
  [event]
  (when (contains?  interesting-events (:method event))
    (send-notification (:source event))))

(defn check-resources 
  ""
  []
  (transaction
    (with-query-results rs 
      	["SELECT distinct(source), method, now() AS now FROM triggers WHERE logdate >= ?::timestamp"  @last-event]
      (when (first rs)
        (dosync (ref-set last-event (str (:now (first rs))))))
      (dorun (map #(process-interesting-events %) (into (hash-set) (map  rs)))))))

(defn  )

(defn run-event-loop 
  ""
  []
  (with-connection database
    (while (= @server-state :running)
      (check-resources) 
      (Thread/sleep *sleep-time*) 
      )))	


;;;;############################################################################[ Main ]########

(defn -main [& arguments]
  (patch-jabber-library)
  ;; connect to the jabber server
  (init-jabber-connection)
  (join-conference)
  (run-event-loop))

                    
  